### PR TITLE
Merging multi-device branch to main.

### PR DESCRIPTION
**Pull request by benvanik**
**TLDR**: nothing should break, `--iree-hal-target-backends=` is deprecated, use `--iree-hal-target-device=` and appropriate target-specific flags instead.

This reworks the target device concept in the IREE pipeline - in some cases introducing the concept (flow and HAL) and in others replacing placeholder mechanisms around stream affinity. This builds upon prior work that added support for enumerating available devices via the HAL and providing multiple devices to the runtime tools by adding the ability to define devices, allowing for execution and storage resources to be assigned a device, and upgrading passes to support multiple devices. "Multi-device" here means several things and all are accomplished with the same mechanism: a single device that may be one of multiple types (multiple CPU/GPU archs, CPU _or_ GPU, etc), multiple homogeneous devices (4 of the same exact GPUs accessed through the same runtime HAL driver), multiple heterogeneous devices (a CPU and a GPU/NPU/etc), and optional devices (a CPU with some portions offloaded to a GPU/NPU if it's compatible and available at runtime). In this way we can provide cross-compilation/targeting, multi-targeting, and multiple devices with one set of flags, compiler analysis, passes dealing with the devices, and runtime infrastructure.

Early warning: **it's strongly discouraged to use device information prior to codegen** - any pass using such information earlier on is a red flag that will receive pushback. IREE is designed first and foremost as a cross-compiler with multi-targeting at its core and radically changing program behavior near the frontend makes it nearly impossible to have configuration control over the compilation pipeline. Consider specializing on device prior to codegen tantamount to using C preprocessor macros based on operating system or architecture: it means that a problem has not been solved and a workaround has been taken. There are exceptionally few cases that require device information early, and those that do can do so in generic ways that do not disturb the debuggability of the program. For example, far better than preprocessor macros in C++ are function calls and if statements (as we can do in our programs), and even better than that are virtual interfaces (ops that are only lowered to one of multiple implementations later on). That disclaimer out of the way: it's now possible to query device information after the input pipeline (global opt/preprocessing/flow). Upstream will push back against doing so in nearly all cases but it is a useful mechanism for downstream projects.

The key change here is that the `--iree-hal-target-backends=` compiler flag has been deprecated. It continues to work for now with the same behavior as before but usage will shift to the replacement `--iree-hal-target-device=` flag. A single instance of this flag defines a single device within the program and repeated uses of it will define new devices. Devices may be named ("my_device") or anonymous (in which case they will be assigned an ordinal like 0 or 1), and each device may be backed by one or more target devices (Vulkan, local host, HIP, etc). Each target device in the compiler (represented by `IREE::HAL::TargetDevice`) may have any number of backends with various configurations (multiple archs, different deployment formats, etc represented by one or more `IREE::HAL::ExecutableTargetAttr` values).

Example flags:
```sh
# Two devices, one the local host device and the other a Vulkan device:
--iree-hal-target-device=local --iree-hal-target-device=vulkan

# One device selecting between Vulkan if available and otherwise use the local host device:
--iree-hal-target-device=vulkan,local

# Two CUDA devices selected by runtime ordinal; at runtime two --device=
# flags are required to configure both devices:
--iree-hal-target-device=cuda[0] --iree-hal-target-device=cuda[1]

# A fully-defined target specification:
--iree-hal-target-device=#hal.device.target<"cuda", {...}, [#hal.executable.target<...>]>

# Named device for defining a reference by #hal.device.promise<@some_name>:
--iree-hal-target-device=some_name=vulkan
```

The device metadata as specified in the compiler is used to produce enumeration code that executes at runtime and queries the available devices to find the appropriate matches. This means that if the program is compiled to target two CUDA devices then at runtime there must be two CUDA devices specified - the indirection allows for the compiled artifact to work with any two CUDA devices targeted by UUID, device ordinal, etc and not just the first and second CUDA device in the system. E.g. `iree-compile --iree-hal-target-device=cuda[0] --iree-hal-target-device=cuda[1]` and `iree-run-module --device=cuda://UUID_A --device=cuda://UUID_B`. Devices targets in the compiler can now specify the ordinal of the device in order to differentiate between multiple devices at runtime (the `cuda[0]` and `cuda[1]` above indicate the first CUDA device and second CUDA device provided to the runtime).

Major new attributes:
* `#hal.device.promise<@device>` is a reference to a device that will be provided at a later stage. Frontends can use this as a placeholder for devices that are specified on the command line without needing to say what those devices are when exporting.
* `#hal.device.alias<"name">` specifies an `IREE::HAL::TargetDevice` in the compiler (`vulkan`, `local`, `hip`, etc) and expands to a full `#hal.device.target` based on target-specific flags.
* `#hal.device.select<[...]>` controls selection by enumerating each device in turn and matching the first found. 
* `#hal.device.fallback<@other_device>` provides a fallback reference that the device will match if no other device matches. Note that having two devices with the same target will create two copies at runtime - if wanting to use the existing device then the fallback mechanism must be used.
* `#hal.device.affinity<@device>` (and optional queue mask) is used on ops to indicate on which device they should execute.

All of the above flags are just syntactic sugar that add the above attributes to the program IR and it's possible for frontends to insert these attributes or ops directly depending on use-case. In most cases leaving placeholders in the IR such that the exact target can be specified during compilation is ideal: this allows one output from the frontend to be used with any number of targets and configurations. Online compilers, though, may want to bake in their exact configuration and can do so without the need for flags that may lose information. The general flow of the `buildHALDeviceAssignmentPassPipeline`/`iree-opt --iree-hal-device-assignment-pipeline` is:
1. `--iree-hal-target-device=` flags are parsed and a `hal.device.targets` attribute is added to the module.
  * `--iree-hal-device-target=cpu_device=local` becomes `hal.device.targets = [#hal.device.alias<"local"> : !hal.device]`
  * `--iree-hal-device-target=cpu_device=local --iree-hal-device-target=gpu_device=cuda,hip` becomes
  ```mlir
  hal.device.targets = {
    cpu_device = #hal.device.alias<"local"> : !hal.device,
    gpu_device = #hal.device.select<[#hal.device.alias<"cuda"> : !hal.device, #hal.device.alias<"hip"> : !hal.device]> : 
  !hal.device
  }
  ```
2. The `hal.device.targets` attribute (if any) is expanded into `util.global` ops for each device. These globals are initialized with one of the supported attributes which are much later turned into enumeration/selection logic. The above multi-device example becomes:
  ```mlir
  builtin.module attributes {stream.affinity.default = #hal.device.affinity<@cpu_device>} {
    util.global private @cpu_device = #hal.device.alias<"local"> : !hal.device
    util.global private @gpu_device = #hal.device.select<[#hal.device.alias<"cuda"> : !hal.device, #hal.device.alias<"hip"> : !hal.device]> : 
  !hal.device
  }
  ```
3. Any `#hal.device.promise` attributes will be changed to reference the globals with the same name. This allows for retargeting of inputs by letting a frontend specify named devices prior to them having been passed on the command line (or inserted by some other pipeline).
4. Any `#hal.device.alias` attributes are converted to full `#hal.device.target` attributes using the appropriate `IREE::HAL::DeviceTarget` implementation.

Upon completion of the pipeline there are globals initialized with either a specific device target or a selection mechanism to pick between targets. From that point onward devices are a structural part of the program and can be referenced by symbol name via attributes like `#hal.device.affinity`.

Programs are expected to specify the device affinity for all operations either explicitly or implicitly. By default (as today) the first device defined will be used but going forward we will want frontends to start specifying devices. To that end the `flow.tensor.transfer` operation was added to allow a tensor to have a device affinity assigned to it. A new analysis is added that allows all tensors (or stream resources) and ops interacting with them to be queried for which device they should be placed on. For example, a frontend can specify multiple devices be used in a computation by transferring the tensors used:
```mlir
util.func private @my_func(%arg0: tensor<4xi32>) -> tensor<4xi32> {
  %arg0_device_a = flow.tensor.transfer %arg0 : tensor<4xi32> to #hal.device.promise<@device_a>
  %compute_device_a = arith.addi %arg0_device_a, %arg0_device_a : tensor<4xi32>
  %transient_device_b = flow.tensor.transfer %compute_device_a : tensor<4xi32> to #hal.device.promise<@device_b>
  %compute_device_b = arith.muli %transient_device_b, %transient_device_b : tensor<4xi32>
  util.return %compute_device_b : tensor<4xi32>
}
```

To avoid copies there are also ways for frontends to indicate where argument and result tensors are placed. The best way (in that it's most general/powerful) is for the frontends to emit `hal.tensor.import`, `hal.tensor.export`, and `hal.tensor.alias` ops directly as they all now take affinities. When using the default ABI translation pass it's possible to add arg/result attrs to public functions, e.g. `util.func public @my_func(%arg0: tensor<2xi32> {iree.abi.affinity = #hal.device.promise<@device_a>}) -> (tensor<2xi32> {iree.abi.affinity = #hal.device.promise<@device_b>})`. Shorthand is provided to allow specifying an `iree.abi.affinity` on functions themselves for when all arguments and results are placed on the same device.

After the point devices are specified, materialized in the program as globals, and referenced either via the magic default attribute, scoped attributes, or explicit transfer operations most of the mechanics are implementation details of the stream and HAL dialect lowerings. Partitioning, allocation, and scheduling in the stream dialect were always affinity-aware and required only minor tweaks as part of this work while the HAL TODOs for multi-device were implemented by memoizing resources per-device and adding the machinery to enumerate and select devices.

This was reviewed in the following chunks and tested in a roll-up PR #17482:
* https://github.com/iree-org/iree/pull/17915
* https://github.com/iree-org/iree/pull/17917
* https://github.com/iree-org/iree/pull/17916
* https://github.com/iree-org/iree/pull/17918
* https://github.com/iree-org/iree/pull/17919
* https://github.com/iree-org/iree/pull/17920

### CODE REVIEW

File: experimental/regression_suite/shark-test-suite-models/sdxl/test_clip.py
Comment (ScottTodd) [inline experimental/regression_suite/shark-test-suite-models/sdxl/test_clip.py:1]: https://github.com/iree-org/iree/actions/runs/10062279931/job/27815542299?pr=17987#step:10:58
```
INFO     root:benchmark_sdxl_rocm.py:292 VAE Decode Dispatch Count: 247 (golden dispatch count 248)
```

You can "fix" the dispatch count CI results for these tests by editing the golden values in this file: https://github.com/iree-org/iree/blob/d07a309d010981d7ecef6bc8bfc573c559395989/.github/workflows/pkgci_regression_test.yml#L338-L340

248 -> 247
1714 -> 1713

Also looks like VAE benchmarks regressed on mi250 and mi300. Wonder if that's due to this PR or just noise. The benchmarks are relatively stable on datacenter GPUs...

https://github.com/iree-org/iree/actions/runs/10062279931/job/27815541759?pr=17987#step:9:132
https://github.com/iree-org/iree/actions/runs/10062279931/job/27815542299?pr=17987#step:10:131
```
INFO     root:benchmark_sdxl_rocm.py:280 VAE Decode Benchmark Time: 85.1 ms (golden time 74.0 ms)
```

If you want to dig deeper and not just update golden values, test files like this one have the `.mlirbc` source linked near the top and then compile flags listed below. Actually iterating on the benchmarks requires a system with the appropriate hardware.

**Review by ScottTodd — COMMENTED** [review]
I reviewed the individual commits here, so generally LGTM. Just looking now at how much merging this will rock the boat.

What merge strategy are you going for with this PR? Merge commit with CI running on just the merge commit?

**Comment by benvanik** [comment]
Planning on a merge commit so the individual commits are preserved. Ideally there should not be any material changes to user workflows (as the existing `--iree-hal-target-backends=` flag still works) but there may be usage-specific bugs in cases we don't have covered in testing - especially in weirder cases of integrations with custom modules and such (though I tried to ensure no changes there).

File: experimental/regression_suite/shark-test-suite-models/sdxl/test_clip.py
Comment (benvanik) [reply experimental/regression_suite/shark-test-suite-models/sdxl/test_clip.py:1]: I don't see the mlirbc or the compile flags for these in there - where are you looking? (this is what I was griping about on discord last night)

Comment (ScottTodd) [reply experimental/regression_suite/shark-test-suite-models/sdxl/test_clip.py:1]: https://github.com/iree-org/iree/blob/d07a309d010981d7ecef6bc8bfc573c559395989/experimental/regression_suite/shark-test-suite-models/sdxl/test_clip.py#L53-L55

https://github.com/iree-org/iree/blob/d07a309d010981d7ecef6bc8bfc573c559395989/experimental/regression_suite/shark-test-suite-models/sdxl/test_clip.py#L88-L105

Comment (ScottTodd) [reply experimental/regression_suite/shark-test-suite-models/sdxl/test_clip.py:1]: Ah... the benchmarking step reuses some output from the test step? That's sorta nice (ensuring that whatever we benchmark is also tested for correctness), but confusing when flags / files / logs are split across steps.

Here is where a compile command producing those statistics runs:
https://github.com/iree-org/iree/actions/runs/10062279931/job/27815542299?pr=17987#step:7:39

Comment (benvanik) [reply experimental/regression_suite/shark-test-suite-models/sdxl/test_clip.py:1]: maybe having a doc comment in the benchmark that references the source and printing that on failure would help - the benchmark py doesn't mention that this comes from that source AFAICT and needing to track sucks, but doc comment would help readers/maintainers and be an easy way to spit out more useful CI errors

Comment (benvanik) [reply experimental/regression_suite/shark-test-suite-models/sdxl/test_clip.py:1]: oh lol that's confusing, I was trying to figure out what new dispatch was being created because it prints `FAILURE: check 247 == 248: SDXL vae decode dispatch count should not regress` and obviously if it regresses it means 247 is the old and 248 is the new - that can use improvement :P

File: .github/workflows/pkgci_regression_test.yml [hunk 1]
```diff
@@ -356,10 +356,10 @@ jobs:
             --goldentime-rocm-e2e-ms 325.0 \
             --goldentime-rocm-unet-ms 77.0 \
             --goldentime-rocm-clip-ms 15.5 \
-            --goldentime-rocm-vae-ms 74.0 \
-            --goldendispatch-rocm-unet 1714 \
+            --goldentime-rocm-vae-ms 85.2 \
+            --goldendispatch-rocm-unet 1713 \
```
Comment (ScottTodd) [inline .github/workflows/pkgci_regression_test.yml]: Would ideally like to know why this particular program went from 74ms to 85ms, if we trust the benchmarking methodology here enough.

**Review by ScottTodd — APPROVED** [review]

File: .github/workflows/pkgci_regression_test.yml (hunk 1 above)
Comment (benvanik) [reply .github/workflows/pkgci_regression_test.yml]: I am really curious, but also don't believe this should be in our benchmark suite:
image
transform dialect, a bunch of non-default flags all throughout the stack, etc

Comment (ScottTodd) [reply .github/workflows/pkgci_regression_test.yml]: Fair enough. This has been on my mind quite a bit lately. I'd like to strike a good balance between core/upstream sanity (default flags must work, medium sized test suites including XFAILs) and early signal for bleeding edge usage that contributors are iterating on.

I'm trying to rework our CI jobs such that the "core" uses standard runners and baseline configurations then each group contributing to the project (e.g. AMD, ARM, Google) can donate CI runners and add limited sets of optional tests/benchmarks that run on PRs/commits. For heavier tests or workflows needing nonstandard configuration (Android phones, wasm/webgpu-capable web browsers, laptops, etc.) those _might_ be able to find a place in-tree but they should probably live out of tree or in private repos (especially for unreleased hardware ;P).

Comment (benvanik) [reply .github/workflows/pkgci_regression_test.yml]: I'm going to try to figure this out, but I probably won't again - this benchmark stuff has a bad UX and takes way too much effort to deal with to be on presubmit.

Comment (benvanik) [reply .github/workflows/pkgci_regression_test.yml]: RE this being a benchmark - I think that's ok and a good heads up like you say - but there's different tiers for workflows: something using so many non-standard flags carefully crafted to produce a specific result by a downstream project and then requiring that we match within microseconds from any change in the compiler pipeline is not possible. If it's a change detector that's one thing (though still noisy and annoying to daily development), but using all those non-default flags is YMMV and downstream things need to remember that and upstream fixes, etc - we can't have a compiler with an infinite number of potential combinations that never regresses anything.

**Comment by benvanik** [comment]
The sdxl regression is due to the pad-to-intrinsics pass, which is not used in any default pipeline and only specified as a preprocessing pass for that benchmark as a flag and nowhere else in the entire project. The lit tests for it pass. And that's why it's bad to have unused/experimental passes used in critical workloads.

**Review by ScottTodd — APPROVED** [review]

**Comment by yuanfz98** [comment]
@benvanik Would you mind providing an e2d test including iree-run-module for this ? Thanks.

**Comment by benvanik** [comment]
Added as `tools/test/iree-run-module-multi.mlir`.

**Comment by ScottTodd** [comment]
> Added as `tools/test/iree-run-module-multi.mlir`.

Nice, thank you.

**Comment by benvanik** [comment]
mi300 runners are down, and the onnx failures are fixed in a #18039. Since mi250 tests are ok I'm going to merge past the hung mi300 tests.
